stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -107,26 +107,65 @@ end
 openblas_get_config() = strip(unsafe_string(ccall((@blasfunc(openblas_get_config), libblas), Ptr{UInt8}, () )))
 
 """
-    set_num_threads(n)
+    set_num_threads(n::Integer)
+    set_num_threads(::Nothing)
+
+Set the number of threads the BLAS library should use equal to `n::Integer`.
+If `nothing` is passed to this function, julia tries to figure out the optimial number of threads.
+The exact heuristic is an implementation detail.
 
-Set the number of threads the BLAS library should use.
+On exotic variants of `BLAS` this function can fail.
 """
-function set_num_threads(n::Integer)
+function set_num_threads(n::Integer)::Nothing
     blas = vendor()
-    if blas === :openblas
-        return ccall((:openblas_set_num_threads, libblas), Cvoid, (Int32,), n)
-    elseif blas === :openblas64
-        return ccall((:openblas_set_num_threads64_, libblas), Cvoid, (Int32,), n)
+    if blas === :openblas || blas == :openblas64
+        return ccall((@blasfunc(openblas_set_num_threads), libblas), Cvoid, (Cint,), n)
     elseif blas === :mkl
         # MKL may let us set the number of threads in several ways
         return ccall((:MKL_Set_Num_Threads, libblas), Cvoid, (Cint,), n)
-    end
-
-    # OSX BLAS looks at an environment variable
-    @static if Sys.isapple()
+    elseif Sys.isapple()
+        # OSX BLAS looks at an environment variable
         ENV["VECLIB_MAXIMUM_THREADS"] = n
+    else
+        @warn "Failed to set number of BLAS threads." maxlog=1
     end
+    return nothing
+end
+
+_tryparse_env_int(key) = tryparse(Int, get(ENV, key, ""))
+
+function set_num_threads(::Nothing)
+    n = something(
+        _tryparse_env_int("OPENBLAS_NUM_THREADS"),
+        _tryparse_env_int("OMP_NUM_THREADS"),
+        max(1, Sys.CPU_THREADS ÷ 2),
+    )
+    set_num_threads(n)
+end
 
+"""
+    get_num_threads()
+
+Get the number of threads the BLAS library is using.
+
+On exotic variants of `BLAS` this function can fail, which is indicated by returning `nothing`.
+"""
+function get_num_threads()::Union{Int, Nothing}
+    blas = vendor()
```

Comment:
```suggestion
get_num_threads()::Union{Int, Nothing} = _get_num_threads()

function _get_num_threads(; _vendor = vendor, _isapple = Sys.isapple)
    blas = _vendor()
```

Reply:
Mhh we could do this, but is it a desirable tradeoff? It makes the code less clear and does it really give a useful test? Seems a bit like testing for the sake of testing.

Reply:
Yeah, I get it that looks stupid to have "testing for the sake of testing" (and it's not like I'm 100% certain this is the Right Thing™ to do).  But it's plausible that

```julia
key = "VECLIB_MAXIMUM_THREADS"
nt = _tryparse_env_int(key)
```

is "refactored" into

```julia
nt = _tryparse_env_int("VECLIB_MAXIMUM_THREADS")
```

unless everyone is careful enough to notice `$key` is used in `@warn`.  I find it unsatisfactory that the only way to avoid a bug in stdlib is "be careful."

Mocking/monkey-patching during test is a thing in other languages so I think it would have been a nobrainer to test this if Julia had a good monkey-patching support.  Since making `get_num_threads` "patchable" is not difficult, I don't see too much downside in this direction.

Reply:
I added a something in this direction, can you take a look if you like it?